pydantic/json.py

```diff
@@ -53,18 +53,17 @@ def pydantic_encoder(obj: Any) -> Any:
     elif is_dataclass(obj):
         return asdict(obj)
 
-    try:
-        encoder = ENCODERS_BY_TYPE[type(obj)]
-    except KeyError:
-        raise TypeError(f"Object of type '{obj.__class__.__name__}' is not JSON serializable")
+    for base in obj.__class__.__mro__:
+        if base in ENCODERS_BY_TYPE:
+            return ENCODERS_BY_TYPE[base](obj)
     else:
-        return encoder(obj)
+        raise TypeError(f"Object of type '{obj.__class__.__name__}' is not JSON serializable")
 
 
 def custom_pydantic_encoder(type_encoders: Dict[Any, Callable[[Type[Any]], Any]], obj: Any) -> Any:
-    encoder = type_encoders.get(type(obj))
-    if encoder:
-        return encoder(obj)
+    for base in obj.__class__.__mro__:
+        if base in type_encoders:
```

Comment:
For what it's worth, I think it would be slightly faster to do:
```
encoder = type_encoders.get(base)
if encoder is not None:
    return encoder(obj)
```
Since that way you only do one lookup instead of two.

Reply:
Edited based on @samuelcolvin 's suggestion for the same, using KeyError like the prior logic.